### PR TITLE
Pass `PlannedPrivate` from PlanResourceChange to ApplyResourceChange

### LINKED ISSUE #2408: `StateFunc` applied twice for resources under PRC
### What happened?

Resources under PRC have the `StateFunc` applied twice to their resources.
This leads to mangled data in state and perma diffs.

`ApplyResourceChange` applies the `StateFunc`, but the preceding `Diff` also applies it.
Often this is fine because `StateFunc` can be applied multiple times (e.g. normalizing json). But for some operations like calculating a hash it's not.

This causes: https://github.com/pulumi/pulumi-aws/issues/4446

### Example

- https://github.com/pulumi/pulumi-aws/issues/4446
- Repro: https://github.com/pulumi/pulumi-terraform-bridge/pull/2407

### Output of `pulumi about`

n/a

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by flostadler**
The `PlannedPrivate` field of the PRC response includes metadata that's required for correctly re-assembling the diff.
For example, it includes `NewExtra` which includes the original value for attributes with a `StateFunc`.
Without `PlannedPrivate`, those attributes get saved to state encoded twice. This can lead to mangled data in state and perma-diffs.

Fixes #2408

### CODE REVIEW

File: pkg/tfbridge/provider_test.go [hunk 1]
```diff
@@ -5113,7 +5113,8 @@ func TestPlanResourceChangeUnknowns(t *testing.T) {
 				"setBlockProps":[],
 				"listBlockProps":[],
 				"nestedListBlockProps":[],
-				"maxItemsOneBlockProp":null
+				"maxItemsOneBlockProp":null,
+				"__meta":"{\"_new_extra_shim\":{}}"
```
Comment (flostadler) [inline pkg/tfbridge/provider_test.go:5117]: I wonder if this will cause pain somewhere else? I.e. re-recording of upgrade tests?

**Comment by VenelinMartinov** [comment]
Is the third parameter on CreateContext/UpdateContext the PlannedPrivate in TF? If so, I think we can write cross-tests here similar to https://github.com/pulumi/pulumi-terraform-bridge/blob/a61fbc4e835def1157b96c4870b4282ff299ae10/pkg/tests/cross-tests/input_cross_test.go#L13 to make sure that what we pass to the TF provider matches what TF passes to the TF provider.

**Comment by flostadler** [comment]
> Is the third parameter on CreateContext/UpdateContext the PlannedPrivate in TF? If so, I think we can write cross-tests here similar to
> 
> https://github.com/pulumi/pulumi-terraform-bridge/blob/a61fbc4e835def1157b96c4870b4282ff299ae10/pkg/tests/cross-tests/input_cross_test.go#L13
> 
> to make sure that what we pass to the TF provider matches what TF passes to the TF provider.

Sadly it's not :(

**Review by iwahbe — COMMENTED** [review]
It's hard for me to review, just because it's hard to know what the correct behavior is. 

I'd leave this for @VenelinMartinov and @t0yv0 unless I need to spin up on this.

**Comment by VenelinMartinov** [comment]
>Sadly it's not :(

Where is the PlannedPrivate picked up on the TF side? Is there a way for us to hook into that and write a cross-test for that? That'd give us more confidence that we are matching the TF behaviour here.

Don't want to block the PR on this since the changes look sensible but it'd be nice if we can be confident here.

**Comment by VenelinMartinov** [comment]
How does the TF provider access the PlannedPrivate?

**Comment by t0yv0** [comment]
I was working with this on Florian a bit this morning, I can review. It looked safe enough but we were exploring how it's tangled to timeouts and there's a little bit of residual uncertainty I have around how TF handles timeout etc encoding in  PlannedPrivate as well as whether we're doing a good job of mapping planned private to Pulumi protocol. It's a bit subtle.

**Comment by flostadler** [comment]
@VenelinMartinov
What we're interested in is this bit here: https://github.com/hashicorp/terraform-plugin-sdk/blob/527cfe4ccf1738d4efcfd6006bbd2068f4a35a06/helper/schema/grpc_provider.go#L1111-L1122

It's reconstructing the `NewExtra` field of the `ResourceAttrDiff` using data in `PlannedPrivate`.
`NewExtra` contains the original value the attribute was set to before the `StateFunc` was applied (even though the name `NewExtra` is misleading).

The resource access methods (e.g. `schema.ResourceData::GetOk`) either return the value in state or `NewExtra` if it's set.
Right now the regression test is asserting what's stored in state.
Can we craft a cross-test out of this?

**Comment by VenelinMartinov** [comment]
Maybe not without bigger changes - this property seems to be hidden from the provider too and only accessible from within the plugin-sdk. The provider only sees the .Get result change depending on if this is set or not.

Is this correct? If so, then we'd need to add a hook in the plugin-sdk for this which is certainly out of scope here.

**Comment by flostadler** [comment]
@VenelinMartinov Yeah you're understanding is correct. It's hidden in the `ResourceData` and the `.Get` result changes are the only thing providers see.

**Comment by t0yv0** [comment]
Cross-test would be very helpful if just to debug the TF side of things.

**Comment by iwahbe** [comment]
Why couldn't we write a cross test that asserts on the results of .Get? We don't need a cross-test to assert the internals, only the provider visible section.

**Comment by VenelinMartinov** [comment]
Yeah, that bit is possible - a test like https://github.com/pulumi/pulumi-terraform-bridge/blob/a61fbc4e835def1157b96c4870b4282ff299ae10/pkg/tests/cross-tests/input_cross_test.go#L13 but with the StateFunc and the modifications as in the existing test.

Unfortunately that wouldn't cover the full contents of the `PlannedPrivate` field.

**Comment by flostadler** [comment]
@t0yv0 @VenelinMartinov I added a cross test, please have a look

**Pull request merged**